File: vigenere.py
#!/rsr/bin/env python3

from sys import argv


def main():
    # check command line arguments
    if len(argv) != 2:
        print("usage: python vigenere.py keyword")
        exit(1)
    elif not argv[1].isalpha():
        print("key must be alphabetical only")
        exit(1)

    # set-up variables for later use
    k = argv[1]  # keyword
    pos = 0  # position indicator for iterating through keyword
    c = []  # list for adding encrypted letters

    # prompt user for plaintext string
    p = input("plaintext: ")

    # encrypt plaintext
    for l in p:
        # encrypt letter l with k[index % len(k)] where index counts +1 with each runthrough of the loop
        c.append(encrypt(l, k[pos % len(k)]))
        if l.isalpha():
            pos += 1

    print("ciphertext: {}".format("".join(c)))


def encrypt(letter, key):
    # the key is not case sensitive
    key = (ord(key.upper()) - 65)

    if letter.isalpha():
        if letter.isupper():
            # 65
            letter = (((ord(letter) - 65) + key) % 26) + 65
            return chr(letter)
        elif letter.islower():
            # 97
            letter = (((ord(letter) - 97) + key) % 26) + 97
            return chr(letter)
    else:
        return letter


if __name__ == "__main__":
    main()

File: vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int shift(char c);

int main(int argc, string argv[])
{
    //check command line inputs
    if (argc != 2)
    {
        printf("Usage: ./vigenere keyword\n");
        return 1;
    }
    else
    {
        int len = strlen(argv[1]); //store length of key in variable
        char keyword[len]; //create array to store the keyword
        //key should be alpha only
        for (int i = 0; i < len; i++)
        {
            if (isalpha(argv[1][i]) == 0)
            {
                printf("Usage: ./vigenere keyword\n");
                return 1;
            }
            else
            {
                keyword[i] = tolower(argv[1][i]);
            }
        }

        //collect plaintext string from user
        string plain = get_string("plaintext: \n");
        //declare cipher array of chars for later use; same size as plaintext string
        char cipher[strlen(plain) + 1];

        //iterate over each character in plaintext & modify each letter by the key
        //declare counter variable j to cycle through keyword
        int j = 0;
        for (int i = 0, plen = strlen(plain); i < plen; i++)
        {
            int key = shift(keyword[j]);
            if (isupper(plain[i]) != 0)
            {
                cipher[i] = (char)(((((int)plain[i] - 65) + key) % 26) + 65);
                j++;
            }
            else if (islower(plain[i]) != 0)
            {
                cipher[i] = (char)(((((int)plain[i] - 97) + key) % 26) + 97);
                j++;
            }
            else
            {
                cipher[i] = plain[i];
            }
            //if j is the same length as the keyword reset it to 0
            if (j == strlen(keyword))
            {
                j = 0;
            }
        }
        cipher[(int)strlen(plain)] = '\0';
        printf("ciphertext: %s\n", cipher);
    }
}

//basic shift function to convert lowercsase char key to alphabetical int
int shift (char c)
{
    c = (int)c - 97;
    return c;
}

